plugins/functions/string.c

```diff
@@ -294,6 +304,165 @@ func_string_tokenize(grn_ctx *ctx, int n_args, grn_obj **args,
+  }
+
+  target = args[0];
+  pattern = args[1];
+  capture = args[2];
+  if (n_args == 4) {
+    default_value = args[3];
+  } else {
+    default_value = grn_plugin_proc_alloc(ctx, user_data, target->header.domain, 0);
+    if (!default_value) {
+      return NULL;
+    }
+
+    GRN_BULK_REWIND(default_value);
+  }
+
+  if (!grn_obj_is_text_family_bulk(ctx, target)) {
+    grn_obj inspected;
+
+    GRN_TEXT_INIT(&inspected, 0);
+    grn_inspect(ctx, &inspected, target);
+    GRN_PLUGIN_ERROR(ctx, GRN_INVALID_ARGUMENT,
+                     "[regex_slice][target] must be a text bulk: <%.*s>",
+                     (int)GRN_TEXT_LEN(&inspected),
+                     GRN_TEXT_VALUE(&inspected));
+    GRN_OBJ_FIN(ctx, &inspected);
+    return NULL;
+  }
+
+  if (!grn_obj_is_text_family_bulk(ctx, capture) && !grn_obj_is_number_family_bulk(ctx, capture)) {
+    grn_obj inspected;
+
+    GRN_TEXT_INIT(&inspected, 0);
+    grn_inspect(ctx, &inspected, capture);
+    GRN_PLUGIN_ERROR(ctx, GRN_INVALID_ARGUMENT,
+                     "[regex_slice][capture] must be a text or number bulk: <%.*s>",
+                     (int)GRN_TEXT_LEN(&inspected),
+                     GRN_TEXT_VALUE(&inspected));
+    GRN_OBJ_FIN(ctx, &inspected);
+    return NULL;
+  }
+
+  //TODO: should cache
+  regex = grn_onigmo_new(ctx,
+                         GRN_TEXT_VALUE(pattern),
+                         GRN_TEXT_LEN(pattern),
+                         GRN_ONIGMO_OPTION_DEFAULT,
+                         GRN_ONIGMO_SYNTAX_DEFAULT,
+                         "[regex_slice][regexp]");
+
+  if (!regex) {
+    return NULL;
+  }
+
+  region = onig_region_new();
+
+  target_string = GRN_TEXT_VALUE(target);
+  target_length = GRN_TEXT_LEN(target);
+
+  //Cannot use normalized string. 
+  //The matching parts of original string cannot be inferred from the matching parts of the normalized string.
+  position = onig_search(regex,
+                         target_string,
+                         target_string + target_length,
+                         target_string,
+                         target_string + target_length,
+                         region,
+                         0);
+
+  if (grn_obj_is_text_family_bulk(ctx, capture)) {
+    capture_name = GRN_TEXT_VALUE(capture);
+    capture_name_length = GRN_TEXT_LEN(capture);
+    capture_num = onig_name_to_backref_number(regex,
+                                              capture_name,
+                                              capture_name + capture_name_length,
+                                              region);
+  } else if (grn_obj_is_number_family_bulk(ctx, capture)) {
+    capture_num = grn_plugin_proc_get_value_int64(ctx,
+                                                  capture,
+                                                  0,
+                                                  "[regex_slice][capture_num]");
+  }
+
+  if (position >= 0 && capture_num >= 0 && region->num_regs > capture_num) {
```

Comment:
I want to put the variable paying attention to left hand side rather than right hand side:

```suggestion
  if (position >= 0 && capture_num >= 0 && capture_num < region->num_regs) {
```

We pay attention `capture_num` rather than `region->num_reges` here.